hydra-node/src/Hydra/Network/Reliability.hs

```diff
@@ -0,0 +1,294 @@
+{-# LANGUAGE TupleSections #-}
+{-# OPTIONS_GHC -Wno-orphans #-}
+
+-- | A `Network` layer that guarantees delivery of `msg` in order even in the
+-- face of transient connection failures.
+--
+-- This layer implements an algorithm based on _vector clocks_, loosely inspired from
+-- /Reliable consistent broadcast/ algorithms with FIFO ordering as presented in
+-- [Introduction to Reliable and Secure Distributed
+-- Programming](https://www.distributedprogramming.net), ch. 3.9, by Cachin et al.
+--
+-- Each node maintains a vector of monotonically increasing integer
+-- indices denoting the index of the last message known (sent or received) for
+-- each peer, where a peer is identified a `Party`, which is updated upon
+-- sending and receiving messages, and is sent with each message.
+--
+-- The basic algorithm is simple:
+--
+--   * When a message is sent, the index of the current node's party is incremented,
+--
+--   * When a message is received:
+--
+--       * It is discarded if the index for the sender's party in the message is
+--         not exactly one more than the latest known index,
+--
+--       * If our own party's index as broadcasted by the sender is lower than our
+--         latest known index, we resend all the "missing" messages.
+--
+-- As shown by the signature of the `withReliability` function, this layer
+-- relies on an authentication layer providing `Authenticated` messages in order
+-- to securely identify senders, and also on `Heartbeat` messages in order to
+-- provide some "liveness".
+--
+-- `Heartbeat` messages are critical in order to /signal/ peers our current view
+-- of the world, because it could be the case that we don't have any network
+-- message to send which leads to head being stalled. `Ping` messages are
+-- treated specially, both when receiving and sending:
+--
+--   * When sending a `Ping`, we /don't increment/ our local message counter
+--     before broadcasting it,
+--
+--   * Conversely, when receiving a `Ping`, we don't update the peer's message
+--     counter but only take into account their view of /our/ counter in order
+--     to compute whether or not to resend previous messages.
```

Comment:
Because the fact we can receive both "meaningful" messages (`Data`) and "heartbeat" messages (`Ping`) is important to detect when to resend messages or not. In essence, we resend messages only when we detect the peer is lagging behing *and* it is quiescent, eg. does not make more progress, which is manifested by `Ping`s which don't change the message sequence number.